### ISSUE TITLE
add native lua support for semantic mediawiki

### ISSUE THREAD

**Issue by oetterer** [SemanticMediaWiki/SemanticMediaWiki#1969]
### Setup and configuration
- SMW version: 2.5@dev
- MW version: 1.27.1
- PHP version: 5.6
- DB (MySQL etc.): mysql 5.5
### Proposal

I would like to suggest a native lua support for semantic mediawiki.
### Motivation

For projects using the [Scribunto](https://www.mediawiki.org/wiki/Extension:Scribunto) Extension the only way of accessing semantic data is by calling the parser function #ask via frame. Besides the fact that this causes a performance issue, to make the resulting semantic data available inside the lua module, one has to do some serious string parsing. Same goes when storing data via #set or #subobject, albeit the string parsing is not necessary.

To streamline this and make it possible for users to directly access smw's data inside lua modules, I would like to suggest creating a smw.lua module that returns the result of an #ask query directly as a table when called inside other lua modules.

Same goes for set and suboject: let users call a smw.lua function to directly store data, w/o having to rely on the slower frame:callParserFunction().
### Notes on existing solutions
- There is the way of using callParserFunction. This is a viable, if not slower way to store data. When it comes to retrieving data, one has to find a way to create a lua data table from a result string. This is tedious and prone to error
- [Extension: Semantic Scribunto](https://www.mediawiki.org/wiki/Extension:Semantic_Scribunto) allows to do an ask query and returns a json formatted string. With Scribunto >= 1.25 this can be parsed into a table. However, this extension relies on open read access to the api (see [Issue#8](https://github.com/SemanticMediaWiki/SemanticScribunto/issues/8) and provides no solution to write property or subobject data.

**Comment by mwjames** [comment]
> I would like to suggest a native lua support for semantic mediawiki.

Lua (or for that matter Scribunto) are not essential for the working of SMW-core hence creating an interdependency on another MediaWiki extension will unlikely met with a favourable support as it increases maintenance cost, technical dept, and may cause unanticipated breakages [0]. Instead, Extension: Semantic Scribunto has been created with those users in mind that want a more tighter integration with Scribunto and SMW without making SMW-core directly depend on it.

In general, functionality that is not absolutely necessary to the overall mission of the software (which is to provide the means of storing and retrieving data in a structured way) is expected to be provided by add-on extensions that can be maintained independently and define necessary dependency boundaries.

> relies on open read access to the api (see Issue#8 and provides no solution to write property or subobject data.

Being addressed by https://github.com/SemanticMediaWiki/SemanticScribunto/pull/9.

> Same goes when storing data via #set or #subobject, albeit the string parsing is not necessary.

If such functionality is desired then Extension: Semantic Scribunto will most likely the place to provide support for such if someone cares to implement them. A pull request that is to implement such functionality are expected to include unit tests.

[0] https://phabricator.wikimedia.org/T148012

**Comment by kghbln** [comment]
@oetterer There is some big progress at [Semantic Scribunto](https://github.com/SemanticMediaWiki/SemanticScribunto/pull/9). You should give it a show and see if this works for you?

> If such functionality is desired then Extension: Semantic Scribunto will most likely the place to provide support for such if someone cares to implement them.

I also think that such functionality should be provided by a separate extension.

**Comment by oetterer** [comment]
Thanks for your input and sorry for the delayed response.

First of all let me explain some things: I started working on this a few month ago and only recently learned about the semantic scribunto extension. I was excited at first until I saw that "experimental" and read about the mentioned issue. So I continued on direct smw integration. However given the new infos on this discussion I agree to aim for semantic scribunto integration.

Before I close this and the coressponding pull request: @mwjames did you by chance get a look at the code? I would like to integrate the functionalities for #info, #set, #subobject and #ask into semantic scribunto (including unit tests) and would value your opinion before I start.

**Comment by mwjames** [comment]
>Thanks for your input and sorry for the delayed response.

Don't worry.

>  I started working on this a few month ago and only recently learned about the semantic scribunto extension. I was excited at first until I saw that "experimental" state and read about the mentioned issue. 

I just recently moved the base classes to that new extension in hopes that some people might find it usable.

> I would like to integrate the functionalities for #info, #set, #subobject and #ask into semantic scribunto (including unit tests) and would value your opinion before I start.

I gave it a quick look (really quick) but some preliminary recommendations are:

- Lua namespace is `mw.smw` (just moved from the `mw.ext.smw`, https://github.com/SemanticMediaWiki/SemanticScribunto/pull/15)
- PHP related namespace is `SMW\Scribunto`
- I would create a separate PR for each new function making review and testing much more easier then having everything crammed in one PR
- I'd like to avoid a reference to `\ParserHooks\FunctionRunner` otherwise one would need an explicit dependency on that package because it is explicit used by SMW but not exposed to the outside  
- `preg_match('~^.*smw-highlighter.*smwttcontent.*>` seems a bit heavy handed as it contains an implicit/internal reference context
- What is the advantage from `ask` over `mw.smw.getQueryResult`? I'm asking because it copies quite some internal processing from the original `#ask` implementation.
